File: java/01-beginner/05-arrays/ArrayBasics.java
import java.util.Arrays;

/**
 * LESSON: Arrays
 * Arrays store multiple values of the same type in a fixed-size container.
 */
public class ArrayBasics {
    public static void main(String[] args) {
        // ===== DECLARING AND INITIALIZING =====
        int[] numbers = new int[5]; // creates array of 5 zeros
        numbers[0] = 10;
        numbers[1] = 20;
        numbers[2] = 30;
        numbers[3] = 40;
        numbers[4] = 50;

        // Shorthand initialization
        String[] colors = {"Red", "Green", "Blue", "Yellow"};
        double[] prices = new double[]{9.99, 19.99, 29.99};

        // ===== ACCESSING ELEMENTS =====
        System.out.println("First number: " + numbers[0]);
        System.out.println("Last color: " + colors[colors.length - 1]);
        System.out.println("Array length: " + numbers.length);

        // ===== ITERATING =====
        System.out.println("\n--- For Loop ---");
        for (int i = 0; i < numbers.length; i++) {
            System.out.println("numbers[" + i + "] = " + numbers[i]);
        }

        System.out.println("\n--- For-Each Loop ---");
        for (String color : colors) {
            System.out.println("Color: " + color);
        }

        // ===== USEFUL ARRAYS METHODS =====
        int[] unsorted = {64, 25, 12, 22, 11};

        System.out.println("\nBefore sort: " + Arrays.toString(unsorted));
        Arrays.sort(unsorted);
        System.out.println("After sort:  " + Arrays.toString(unsorted));

        // Copy array
        int[] copy = Arrays.copyOf(unsorted, unsorted.length);
        System.out.println("Copy: " + Arrays.toString(copy));

        // Fill array
        int[] filled = new int[5];
        Arrays.fill(filled, 7);
        System.out.println("Filled: " + Arrays.toString(filled));

        // Binary search (array must be sorted)
        int index = Arrays.binarySearch(unsorted, 22);
        System.out.println("22 found at index: " + index);

        // Compare arrays
        System.out.println("Arrays equal? " + Arrays.equals(unsorted, copy));
    }
}

File: cpp-practice/01-beginner/05-arrays/array_basics.cpp
/**
 * LESSON: Arrays (C-style and std::array)
 */
#include <iostream>
#include <array>     // std::array (C++11)
#include <algorithm> // sort, fill, etc.

int main() {
    // ===== C-STYLE ARRAYS =====
    std::cout << "--- C-Style Arrays ---\n";
    int numbers[5] = {10, 20, 30, 40, 50};
    std::string colors[] = {"Red", "Green", "Blue"}; // size inferred

    std::cout << "First: " << numbers[0] << "\n";
    std::cout << "Last: " << numbers[4] << "\n";
    std::cout << "Size: " << sizeof(numbers) / sizeof(numbers[0]) << "\n";

    // Iterating
    std::cout << "\n--- Iteration ---\n";
    for (int i = 0; i < 5; i++) {
        std::cout << "numbers[" << i << "] = " << numbers[i] << "\n";
    }

    // Range-based for
    for (const auto& color : colors) {
        std::cout << "Color: " << color << "\n";
    }

    // ===== STD::ARRAY (C++11 - preferred) =====
    std::cout << "\n--- std::array ---\n";
    std::array<int, 5> arr = {64, 25, 12, 22, 11};

    std::cout << "Size: " << arr.size() << "\n";
    std::cout << "Front: " << arr.front() << "\n";
    std::cout << "Back: " << arr.back() << "\n";
    std::cout << "At(2): " << arr.at(2) << "\n"; // bounds-checked

    // Sorting
    std::sort(arr.begin(), arr.end());
    std::cout << "Sorted: ";
    for (int n : arr) std::cout << n << " ";
    std::cout << "\n";

    // Fill
    std::array<int, 4> filled;
    filled.fill(7);
    std::cout << "Filled: ";
    for (int n : filled) std::cout << n << " ";
    std::cout << "\n";

    // Empty check
    std::cout << "Empty? " << std::boolalpha << arr.empty() << "\n";

    // ===== 2D ARRAY =====
    std::cout << "\n--- 2D Array ---\n";
    int matrix[3][3] = {
        {1, 2, 3},
        {4, 5, 6},
        {7, 8, 9}
    };

    for (int i = 0; i < 3; i++) {
        for (int j = 0; j < 3; j++) {
            std::cout << matrix[i][j] << "\t";
        }
        std::cout << "\n";
    }

    return 0;
}
